common/setups/hybrid/util.py

```diff
@@ -0,0 +1,399 @@
+    "GmmDataInput",
+    "GmmInitArgs",
+    "GmmMonophoneArgs",
+    "GmmTriphoneArgs",
+    "GmmVtlnArgs",
+    "GmmSatArgs",
+    "GmmVtlnSatArgs",
+    "NnArgs",
+]
+
+import i6_core.meta as meta
+
+
+class GmmDataInput:
+    def __init__(
+        self,
+        corpus_object: meta.CorpusObject = None,
+        lexicon: dict = None,
+        lm: dict = None,
+        concurrent: int = 10,
+    ):
+        """
+        :param corpus_object: corpus_file: Path, audio_dir: Path, audio_format: str, duration: float
+        :param lexicon: file: Path, normalize_pronunciation: bool
+        :param lm: file: Path, type: str, scale: float
+        :param concurrent: concurrency for gmm hmm pipeline
+        """
+        self.corpus_object = corpus_object
+        self.lexicon = lexicon
+        self.lm = lm
+        self.concurrent = concurrent
+
+
+"""
+when getting an zero weights error, set:
+
+extra_config = sprint.SprintConfig()
+extra_config.allow_zero_weights = True
+{accumulate,split,align}_extra_args = {'extra_config': extra_config}
+
+'{accumulate,split,align}_extra_rqmt': {'mem': 10, 'time': 8},
+
+vtln align time = 8
+
+if not using the run function -> name and corpus almost always to be added
+"""
+
+
+class GmmInitArgs:
+    def __init__(
+        self,
+        costa_args: dict = None,
+        am_args: dict = None,
+        feature_extraction_args: dict = None,
+    ):
+        """
+        ##################################################
+        :param costa_args"
+        {
+          'eval_recordings': True,
+          'eval_lm': True
+        }
+        ##################################################
+        :param am_args:
+        {
+          state_tying="monophone",
+          states_per_phone=3,
+          state_repetitions=1,
+          across_word_model=True,
+          early_recombination=False,
+          tdp_scale=1.0,
+          tdp_transition=(3.0, 0.0, 30.0, 0.0),  # loop, forward, skip, exit
+          tdp_silence=(0.0, 3.0, "infinity", 20.0),
+          tying_type="global",
+          nonword_phones="",
+          tdp_nonword=(0.0, 3.0, "infinity", 6.0)  # only used when tying_type = global-and-nonword
+        }
+        ##################################################
+        :param feature_extraction_args:
+          'mfcc':
+          {
+            'num_deriv': 2,
+            'num_features': None,  # confusing name: number of max features, above number -> clipped
+            'mfcc_options':
+            {
+              'warping_function': "mel",
+              'filter_width': 268.258,  # 80
+              'normalize': True,
+              'normalization_options': None,
+              'without_samples': False,
+              'samples_options': {'audio_format': "wav",
+                                  'dc_detection': True,},
+              'cepstrum_options': {'normalize': False,
+                                   'outputs': 16,
+                                   'add_epsilon': False,},
```

Comment:
I don't really like these hanging indents.
(But it's just the documentation, so maybe doesn't matter...)